File: packages/jwies-qt-client/jwies_qt_client/cards.py
"""Card codes to SVG element ids.

The client deliberately does not depend on ``jwies-core``: it only needs to
turn a wire code like ``"10S"`` into the element id ``"10_spade"`` used by
svg-cards.svg. Keeping this table here means the client carries no rules
engine, which is also why it can never disagree with the server about them.
"""

from __future__ import annotations

from typing import Final

__all__ = ["CARD_BACK", "card_label", "suit_label", "svg_element_id"]

CARD_BACK: Final = "back"

_SUIT_SVG: Final = {"C": "club", "D": "diamond", "H": "heart", "S": "spade"}
_RANK_SVG: Final = {
    "2": "2",
    "3": "3",
    "4": "4",
    "5": "5",
    "6": "6",
    "7": "7",
    "8": "8",
    "9": "9",
    "10": "10",
    "J": "jack",
    "Q": "queen",
    "K": "king",
    "A": "1",
}

_SUIT_NL: Final = {"C": "klaveren", "D": "koeken", "H": "harten", "S": "schoppen"}
_RANK_NL: Final = {
    "2": "twee",
    "3": "drie",
    "4": "vier",
    "5": "vijf",
    "6": "zes",
    "7": "zeven",
    "8": "acht",
    "9": "negen",
    "10": "tien",
    "J": "boer",
    "Q": "dame",
    "K": "heer",
    "A": "aas",
}


def svg_element_id(code: str) -> str:
    """``"10S"`` -> ``"10_spade"``. Returns ``"back"`` unchanged."""
    if code == CARD_BACK:
        return CARD_BACK
    rank, suit = code[:-1], code[-1]
    return f"{_RANK_SVG[rank]}_{_SUIT_SVG[suit]}"


def card_label(code: str) -> str:
    """A Dutch name for a card, used as an accessible label."""
    if code == CARD_BACK:
        return "gedekte kaart"
    rank, suit = code[:-1], code[-1]
    return f"{_SUIT_NL[suit]} {_RANK_NL[rank]}"


def suit_label(suit: str | None) -> str:
    return _SUIT_NL.get(suit or "", "zonder troef")

File: jwies-web-client/jwies_web_client/static/js/cards.js
// Kaarten tekenen uit svg-cards.svg.
//
// Dat bestand is een grote plaat met 53 <g>-groepen zonder eigen transform.
// Externe <use href="bestand.svg#id"> werkt niet betrouwbaar in alle browsers,
// dus halen we de plaat een keer op, verbergen we ze, en knippen we per kaart
// exact de juiste rechthoek uit via getBBox() + viewBox.

const SUIT_SVG = { C: "club", D: "diamond", H: "heart", S: "spade" };
const RANK_SVG = {
  2: "2", 3: "3", 4: "4", 5: "5", 6: "6", 7: "7", 8: "8", 9: "9", 10: "10",
  J: "jack", Q: "queen", K: "king", A: "1",
};

const boxes = new Map();
let ready = false;

export function svgElementId(code) {
  const rank = code.slice(0, -1);
  const suit = code.slice(-1);
  return `${RANK_SVG[rank]}_${SUIT_SVG[suit]}`;
}

export async function loadDeck() {
  if (ready) return;
  const response = await fetch("/assets/svg-cards.svg");
  const markup = await response.text();
  const host = document.getElementById("card-sprite");
  // Zero-sized met overflow:hidden, niet display:none - getBBox() heeft een
  // daadwerkelijk gerenderd element nodig om te kunnen meten.
  host.setAttribute(
    "style",
    "position:absolute;width:0;height:0;overflow:hidden;",
  );
  host.innerHTML = markup;
  ready = true;
}

function boxOf(elementId) {
  if (boxes.has(elementId)) return boxes.get(elementId);
  const node = document.getElementById(elementId);
  if (!node) return null;
  const box = node.getBBox();
  const value = `${box.x} ${box.y} ${box.width} ${box.height}`;
  boxes.set(elementId, value);
  return value;
}

/** Bouwt een <svg> dat exact een kaart uit de plaat toont. */
export function cardElement(code, { faceDown = false } = {}) {
  const elementId = faceDown ? "back" : svgElementId(code);
  const svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  svg.classList.add("card");
  svg.dataset.card = faceDown ? "back" : code;
  const viewBox = boxOf(elementId);
  if (viewBox) svg.setAttribute("viewBox", viewBox);
  svg.setAttribute("preserveAspectRatio", "xMidYMid meet");
  if (!faceDown) svg.setAttribute("aria-label", cardLabel(code));

  const use = document.createElementNS("http://www.w3.org/2000/svg", "use");
  use.setAttribute("href", `#${elementId}`);
  svg.appendChild(use);
  return svg;
}

const RANK_NL = {
  2: "twee", 3: "drie", 4: "vier", 5: "vijf", 6: "zes", 7: "zeven",
  8: "acht", 9: "negen", 10: "tien", J: "boer", Q: "dame", K: "heer", A: "aas",
};
const SUIT_NL = { C: "klaveren", D: "koeken", H: "harten", S: "schoppen" };

export function cardLabel(code) {
  const rank = code.slice(0, -1);
  const suit = code.slice(-1);
  return `${SUIT_NL[suit]} ${RANK_NL[rank]}`;
}

export function suitLabel(suit) {
  return SUIT_NL[suit] ?? "zonder troef";
}

export const SUIT_SYMBOL = { C: "♣", D: "♦", H: "♥", S: "♠" };
